src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/poptimizer/simple/ReduceVocRewritingJoinPlanOptimizerImpl.java

```diff
@@ -0,0 +1,144 @@
+                final PriorityQueue<PhysicalPlanWithStatistics> orderedCandidateSubPlans = new PriorityQueue<>(orderBasedOnCard);
+                PhysicalPlanWithStatistics nextSelectedSubPlan = firstSubPlan;
+                PhysicalPlan currentJoinPlan = nextSelectedSubPlan.plan;
+
+                while ( !subPlansWithStatistics.isEmpty() ){
+                    // Identify subplans that have join variables with the selected subplan as new candidateSubPlans.
+                    List<PhysicalPlanWithStatistics> tmpCandidateSubPlans = getSubPlansContainVars( currentJoinPlan.getExpectedVariables(), subPlansWithStatistics );
+                    if( tmpCandidateSubPlans.isEmpty() ){
+                        tmpCandidateSubPlans = subPlansWithStatistics;
+                    }
+
+                    // Keep candidate SubPlans that use the same vocabulary mapping with the previous selected SubPlan.
+                    // If 'candidateSubPlans' is empty, use all subplans in 'tmpCandidateSubPlans' as candidate subplans.
+                    List<PhysicalPlanWithStatistics> candidateSubPlans = getSubPlansWithSameVoc( nextSelectedSubPlan.getVocabularyMappings(), tmpCandidateSubPlans);
+                    if( candidateSubPlans.isEmpty() ){
+                        candidateSubPlans = tmpCandidateSubPlans;
+                    }
+
+                    // The candidateSubPlans are added to the ordered list 'orderedCandidateSubPlans'
+                    orderedCandidateSubPlans.addAll(candidateSubPlans);
+
+                    // select the first element (with the lowest cardinality) from orderedCandidateSubPlans as the nextSelectedSubPlan
+                    // Remove the selected subPlan from subPlansWithStatistics and clear orderedCandidateSubPlans for the next loop
+                    nextSelectedSubPlan = orderedCandidateSubPlans.poll();
+                    subPlansWithStatistics.remove(nextSelectedSubPlan);
+                    orderedCandidateSubPlans.clear();
+
+                    // Construct a physical plan using bind join
+                    currentJoinPlan = constructPlanWithBJ( currentJoinPlan, nextSelectedSubPlan.plan );
+                }
+
+                return currentJoinPlan;
+            }
+        };
+    }
+
+    /**
+     * Iterate through the remaining subplans and selected those that have the same vocabulary mapping in the given list of vocabulary mappings.
+     */
+    protected List<PhysicalPlanWithStatistics> getSubPlansWithSameVoc( final List<VocabularyMapping> vms, final List<PhysicalPlanWithStatistics> subPlansWithStatistics ) {
+        final List<PhysicalPlanWithStatistics> subPlansWithSameVoc = new ArrayList<>();
+        for ( final PhysicalPlanWithStatistics subplan: subPlansWithStatistics ) {
+            final List<VocabularyMapping> vmsNext = subplan.getVocabularyMappings();
+            vms.retainAll(vmsNext);
```

Comment:
This changes `vms` which is not what you want. Instead, can you swap it (`vmsNext.retainAll(vms)`)?